operator/controllers/credential_controller.go

```diff
@@ -0,0 +1,572 @@
+		}
+		kwg := kindWithGroup{kind: obj.Kind, group: gv.Group}
+		resource, err := r.resourceForKind(kwg)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to bind; could not find resource for kind %q: %w", kwg, err)
+		}
+		helmObjectsPolicyRules[i] = rbacv1.PolicyRule{
+			APIGroups:     []string{gv.Group},
+			Resources:     []string{resource.Name},
+			ResourceNames: []string{obj.Name},
+			Verbs:         []string{"get"},
+		}
+	}
+
+	desiredRole := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+			// TODO: add proper labels.
+		},
+		Rules: append(
+			// A rule to allow patching the credential secret.
+			[]rbacv1.PolicyRule{{
+				APIGroups:     []string{""},
+				Resources:     []string{"secrets"},
+				ResourceNames: []string{credentialsSecretName},
+				Verbs:         []string{"get", "patch"},
+			}},
+			// The rules to allow reading the Kubernetes objects the Helm release created.
+			helmObjectsPolicyRules...,
+		),
+	}
+	if err := ctrl.SetControllerReference(credential, desiredRole, r.scheme); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+	}
+
+	currentRole := &rbacv1.Role{}
+	if err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      desiredRole.Name,
+			Namespace: desiredRole.Namespace,
+		},
+		currentRole,
+	); err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+		}
+		if err := r.Create(ctx, desiredRole); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if !equality.Semantic.DeepDerivative(desiredRole.Rules, currentRole.Rules) {
+		if err := r.Update(ctx, desiredRole); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	desiredRoleBinding := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+			// TODO: add proper labels.
+		},
+		Subjects: []rbacv1.Subject{{
+			Kind:      "ServiceAccount",
+			Name:      desiredServiceAccount.Name,
+			Namespace: desiredServiceAccount.Namespace,
+		}},
+		RoleRef: rbacv1.RoleRef{
+			Kind:     "Role",
+			APIGroup: "rbac.authorization.k8s.io",
+			Name:     desiredRole.Name,
+		},
+	}
+	if err := ctrl.SetControllerReference(credential, desiredRoleBinding, r.scheme); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+	}
+
+	currentRoleBinding := &rbacv1.RoleBinding{}
+	if err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      desiredRoleBinding.Name,
+			Namespace: desiredRoleBinding.Namespace,
+		},
+		currentRoleBinding,
+	); err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+		}
+		if err := r.Create(ctx, desiredRoleBinding); err != nil {
```

Comment:
I assume the role binding goes away when the role does? (Or do we need a controller reference here too?)

Reply:
No, the role binding survives a role deletion, but it doesn't survive the credential deletion because of line 302.

Reply:
Though, things get deleted when they are not required anymore starting on line 488.